Introduce reverse view_funcs

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113280
* #113279
* #116308
* #116512
* __->__ #115894

Part 2 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Details:
* Codegen `rev_view_func()` alongside `view_func()`
    * Reverse view_func gives you a "base" from a "view": `rev_view_func(new_view) -> new_base` AKA it plays the original view backwards
* Utilizes the functional inverses defined in `FunctionalInverses.cpp`, passing `InverseReturnMode::AlwaysView`
* Manually implements functional inverses for `narrow()` and `chunk()`
* **NB: Multi-output views now set view_func() / rev_view_func() for each of the output views!**
    * Due to this, the `as_view()` overload that operates on a list of views is scrapped in favor of iteration via codegen

Example codegen in `ADInplaceOrViewTypeN.cpp`:
```cpp
at::Tensor narrow(c10::DispatchKeySet ks, const at::Tensor & self, int64_t dim, c10::SymInt start, c10::SymInt length) {
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::_ops::narrow::redispatch(ks & c10::after_ADInplaceOrView_keyset, self, dim, start, length);
  })();
  std::function<at::Tensor(const at::Tensor&)> func=nullptr;
  std::function<at::Tensor(const at::Tensor&)> rev_func=nullptr;
  if (false || !self.unsafeGetTensorImpl()->support_as_strided() ||
      c10::AutogradState::get_tls_state().get_view_replay_enabled()) {
    func = [=](const at::Tensor& input_base) {
      return at::_ops::narrow::call(input_base, dim, start, length);
    };
    rev_func = [=](const at::Tensor& input_view) {
      // NB: args from narrow() signature are passed along to the inverse
      return at::functionalization::FunctionalInverses::narrow_copy_inverse(self, input_view, at::functionalization::InverseReturnMode::AlwaysView, dim, start, length);
    };
  }
  auto result = as_view(/* base */ self, /* output */ _tmp, /* is_bw_differentiable */ true, /* is_fw_differentiable */ true, /* view_func */ func, /* rev_view_func */ rev_func, /* creation_meta */ InferenceMode::is_enabled() ? CreationMeta::INFERENCE_MODE : (at::GradMode::is_enabled() ? CreationMeta::DEFAULT : CreationMeta::NO_GRAD_MODE));
  return result;
}
```

jbschlosser committed:
Introduce reverse view_funcs

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


TODO:
* Fix the `narrow()` issue
* Add tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


TODO:
* Fix the `narrow()` issue
* Add tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* Fix failing tests for ops that return multiple views (e.g. `split`)
* Add rev view_func chain tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* Fix failing tests for ops that return multiple views (e.g. `split`)
* Add rev view_func chain tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* Fix failing tests for ops that return multiple views (e.g. `split`)
* Add rev view_func chain tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* Add rev view_func chain tests

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* Add OpInfo test stuff

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* Add OpInfo test stuff

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* Add OpInfo test stuff

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


Part 2 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Details:
* Codegen `rev_view_func()` alongside `view_func()`
    * Reverse view_func gives you a "base" from a "view": `rev_view_func(new_view) -> new_base` AKA it plays the original view backwards
* Utilizes the functional inverses defined in `FunctionalInverses.cpp`, passing `InverseReturnMode::AlwaysView`
* **NB: Multi-output views now set view_func() / rev_view_func() for each of the output views!**
    * Due to this, the `as_view()` overload that operates on a list of views is scrapped in favor of iteration via codegen

TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* Add OpInfo test stuff

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


Part 2 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Details:
* Codegen `rev_view_func()` alongside `view_func()`
    * Reverse view_func gives you a "base" from a "view": `rev_view_func(new_view) -> new_base` AKA it plays the original view backwards
* Utilizes the functional inverses defined in `FunctionalInverses.cpp`, passing `InverseReturnMode::AlwaysView`
* Manually implements functional inverses for `narrow()` and `chunk()`
* **NB: Multi-output views now set view_func() / rev_view_func() for each of the output views!**
    * Due to this, the `as_view()` overload that operates on a list of views is scrapped in favor of iteration via codegen

TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* Add OpInfo test stuff

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


Part 2 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Details:
* Codegen `rev_view_func()` alongside `view_func()`
    * Reverse view_func gives you a "base" from a "view": `rev_view_func(new_view) -> new_base` AKA it plays the original view backwards
* Utilizes the functional inverses defined in `FunctionalInverses.cpp`, passing `InverseReturnMode::AlwaysView`
* Manually implements functional inverses for `narrow()` and `chunk()`
* **NB: Multi-output views now set view_func() / rev_view_func() for each of the output views!**
    * Due to this, the `as_view()` overload that operates on a list of views is scrapped in favor of iteration via codegen

TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* Add OpInfo test stuff

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


Part 2 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Details:
* Codegen `rev_view_func()` alongside `view_func()`
    * Reverse view_func gives you a "base" from a "view": `rev_view_func(new_view) -> new_base` AKA it plays the original view backwards
* Utilizes the functional inverses defined in `FunctionalInverses.cpp`, passing `InverseReturnMode::AlwaysView`
* Manually implements functional inverses for `narrow()` and `chunk()`
* **NB: Multi-output views now set view_func() / rev_view_func() for each of the output views!**
    * Due to this, the `as_view()` overload that operates on a list of views is scrapped in favor of iteration via codegen

TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* Add OpInfo test stuff

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


Part 2 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Details:
* Codegen `rev_view_func()` alongside `view_func()`
    * Reverse view_func gives you a "base" from a "view": `rev_view_func(new_view) -> new_base` AKA it plays the original view backwards
* Utilizes the functional inverses defined in `FunctionalInverses.cpp`, passing `InverseReturnMode::AlwaysView`
* Manually implements functional inverses for `narrow()` and `chunk()`
* **NB: Multi-output views now set view_func() / rev_view_func() for each of the output views!**
    * Due to this, the `as_view()` overload that operates on a list of views is scrapped in favor of iteration via codegen

TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* ~~Add OpInfo tests~~

[ghstack-poisoned]

jbschlosser committed:
Update on "Introduce reverse view_funcs"


Part 2 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Details:
* Codegen `rev_view_func()` alongside `view_func()`
    * Reverse view_func gives you a "base" from a "view": `rev_view_func(new_view) -> new_base` AKA it plays the original view backwards
* Utilizes the functional inverses defined in `FunctionalInverses.cpp`, passing `InverseReturnMode::AlwaysView`
* Manually implements functional inverses for `narrow()` and `chunk()`
* **NB: Multi-output views now set view_func() / rev_view_func() for each of the output views!**
    * Due to this, the `as_view()` overload that operates on a list of views is scrapped in favor of iteration via codegen

TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* ~~Add OpInfo tests~~

[ghstack-poisoned]

albanD commented:
Will let @soulitzer look at this one. Feel free to ping me if you need me!

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -1020,9 +1020,7 @@ def f3(a):
-        # We rely on autograd's view-replay here, and view-replay gives up and uses as_strided
-        # for multi-output views
-        self.assertTrue(all('AsStridedBackward' in str(o.grad_fn) for o in out_test[:3]))
+        self.assertTrue(all('UnbindBackward' in str(o.grad_fn) for o in out_test[:3]))
```
🤔 this might improve perf in some (rare) cases, if we're now avoiding `AsStridedBackward` for multi-output views in all cases.

jbschlosser committed:
Update on "Introduce reverse view_funcs"


Part 2 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Details:
* Codegen `rev_view_func()` alongside `view_func()`
    * Reverse view_func gives you a "base" from a "view": `rev_view_func(new_view) -> new_base` AKA it plays the original view backwards
* Utilizes the functional inverses defined in `FunctionalInverses.cpp`, passing `InverseReturnMode::AlwaysView`
* Manually implements functional inverses for `narrow()` and `chunk()`
* **NB: Multi-output views now set view_func() / rev_view_func() for each of the output views!**
    * Due to this, the `as_view()` overload that operates on a list of views is scrapped in favor of iteration via codegen

TODO:
* ~~Fix the `narrow()` issue~~
* ~~Add reverse view func replay tests~~
* ~~Fix failing tests for ops that return multiple views (e.g. `split`)~~
* ~~Add rev view_func chain tests~~
* ~~Add OpInfo tests~~

[ghstack-poisoned]

soulitzer commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -496,18 +502,23 @@ static Tensor _make_dual(
       return input_base.view(size_vec);
     };
+    rev_func = [=](const at::Tensor& input_view) {
+      return input_view.view(size_vec);
```
Hmm I don't think these actually follow the invariant of rev_view(view(base)) having the same metadata as base. Maybe just put a error for now saying its not supported for now?

jbschlosser replied:
Added an error here and below

soulitzer commented on torch/csrc/autograd/VariableTypeManual.cpp:
```diff
@@ -468,18 +469,23 @@ static Tensor _fw_primal(
       return input_base.view(size_vec);
     };
+    rev_func = [=](const at::Tensor& input_view) {
+      return input_view.view(size_vec);
```
(same for here)

soulitzer commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -528,24 +528,38 @@ static void autogradNotImplementedInplaceOrViewFallbackImpl(
+          false,
+          "Mutating the view ",
+          op_name,
+          " which does not have a derivative implemented is forbidden.");
```
I don't think we would actually want to error here for the rev_view case if we want this view to work for PT2.
Would be nice to update this error message though, since folks can now reach this path without doing in-place over view.

jbschlosser replied:
You're right: for `rev_view_func()` we can reach this path without an in-place over a view. I expect it to (eventually) be hit when attempting to meta-ify a view output of an op that utilizes this fallback. So at the very least, there should be a separate error msg for `rev_view_func()`.

Maybe the error message should mention this scenario specifically? cc @albanD wdyt?

jbschlosser replied:
For posterity: resolved with separate error msg for `rev_view_func()`

soulitzer commented on torch/csrc/autograd/variable.h:
```diff
@@ -332,9 +332,13 @@ struct TORCH_API ViewInfo {
   /// If view_fn has value, we use it to recover views in backward.
   std::function<Variable(const Variable&)> view_fn_;
+  /// Analogue of view_fn but in reverse: given a view -> produce a base that is
+  /// the inverse view of view.
```
nit: I think you mean "analogue of view_fn but in reverse: given a view -> produce the base by applying the inverse view"?

soulitzer commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -93,14 +113,21 @@ ViewInfo ViewInfo::chain(
+        auto error_msg =
+            ("This view is the output of a function that returns multiple views."
+             "Such functions do not allow the output views to be modified inplace."
+             "You should replace the inplace operation by an out-of-place one");
```
We can run into this case now without doing in-place over view right? Maybe we shouldn't error here?

jbschlosser replied:
cc @albanD can I get your thoughts here?

My understanding is that this branch is now only called if:
* We're chaining two views
* Current view has a `view_func()` but parent view does not
* The base tensor does not support `as_strided()`, so we can't cover the gap of the parent view's missing `view_func()`

Note that multiple-view ops like `unbind()` now have `view_func()`s, so the error message is certainly wrong when it mentions the view being an output of a function that returns multiple views.

Can we just error for both `view_func()` / `rev_view_func()` explaining that we can't chain views here because `as_strided()` isn't supported? I expect this to be fairly rare / maybe impossible to hit, as it'd need to be a subclass that doesn't support `as_strided()`, which we should unconditionally generate `view_func()`s for now.

jbschlosser replied:
For posterity: changed error msg to describe the situation literally for both forward / reverse

soulitzer reviewed (commented):
Looking great! Added some small comments. 
btw, could you add the codegen'd output to the description?

jbschlosser commented on aten/src/ATen/templates/FunctionalInverses.h:
```diff
@@ -22,6 +22,10 @@ struct FunctionalInverses {
+// NB: These are not generated! They're manually implemented in the template.
+static at::Tensor chunk_copy_inverse(const at::Tensor & base, const at::Tensor & mutated_view, InverseReturnMode inverse_return_mode, int64_t mutated_view_idx, int chunks, int dim);
+static at::Tensor narrow_copy_inverse(const at::Tensor & base, const at::Tensor & mutated_view, InverseReturnMode inverse_return_mode, int dim, c10::SymInt start, c10::SymInt length);
+
```
> It would be nice though if these are generated as long as ADInplaceOrView kernels are generated.

I agree! But that requires some amount of codegen rework. Currently:
* Functionalization codegen operates on `NativeFunctionsViewGroup`s.
    * There are no `NativeFunctionsViewGroup`s for narrow (see comment [here](https://github.com/pytorch/pytorch/blob/70f3a530d7290d9eb010ad789787bec4dcbfbeb9/torchgen/model.py#L2583-L2585)) or chunk, which is why they have to be defined here.
* ADInplaceOrViewTypeX codegen operates on `NativeFunctionWithDifferentiabilityInfo`s.

It might be possible to clean this up. I'll add a TODO here for re-examining codegen

soulitzer replied:
Sounds good!

jbschlosser committed:
Update on "Introduce reverse view_funcs"


Part 2 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

Details:
* Codegen `rev_view_func()` alongside `view_func()`
    * Reverse view_func gives you a "base" from a "view": `rev_view_func(new_view) -> new_base` AKA it plays the original view backwards
* Utilizes the functional inverses defined in `FunctionalInverses.cpp`, passing `InverseReturnMode::AlwaysView`
* Manually implements functional inverses for `narrow()` and `chunk()`
* **NB: Multi-output views now set view_func() / rev_view_func() for each of the output views!**
    * Due to this, the `as_view()` overload that operates on a list of views is scrapped in favor of iteration via codegen

Example codegen in `ADInplaceOrViewTypeN.cpp`:
```cpp
at::Tensor narrow(c10::DispatchKeySet ks, const at::Tensor & self, int64_t dim, c10::SymInt start, c10::SymInt length) {
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::_ops::narrow::redispatch(ks & c10::after_ADInplaceOrView_keyset, self, dim, start, length);
  })();
  std::function<at::Tensor(const at::Tensor&)> func=nullptr;
  std::function<at::Tensor(const at::Tensor&)> rev_func=nullptr;
  if (false || !self.unsafeGetTensorImpl()->support_as_strided() ||
      c10::AutogradState::get_tls_state().get_view_replay_enabled()) {
    func = [=](const at::Tensor& input_base) {
      return at::_ops::narrow::call(input_base, dim, start, length);
    };
    rev_func = [=](const at::Tensor& input_view) {
      // NB: args from narrow() signature are passed along to the inverse
      return at::functionalization::FunctionalInverses::narrow_copy_inverse(self, input_view, at::functionalization::InverseReturnMode::AlwaysView, dim, start, length);
    };
  }
  auto result = as_view(/* base */ self, /* output */ _tmp, /* is_bw_differentiable */ true, /* is_fw_differentiable */ true, /* view_func */ func, /* rev_view_func */ rev_func, /* creation_meta */ InferenceMode::is_enabled() ? CreationMeta::INFERENCE_MODE : (at::GradMode::is_enabled() ? CreationMeta::DEFAULT : CreationMeta::NO_GRAD_MODE));
  return result;
}
```

[ghstack-poisoned]

soulitzer reviewed (approved):
LGTM

jbschlosser commented:
@pytorchbot merge
